opal/mca/accelerator/rocm/accelerator_rocm_component.c

```diff
@@ -154,28 +158,55 @@ static int accelerator_rocm_component_register(void)
     return OPAL_SUCCESS;
 }
 
+int opal_accelerator_rocm_lazy_init()
+{
+    int err = OPAL_SUCCESS;
+
+    /* Double checked locking to avoid having to
+     * grab locks post lazy-initialization.  */
+    opal_atomic_rmb();
+    if (true == accelerator_rocm_init_complete) {
+        return OPAL_SUCCESS;
+    }
+    OPAL_THREAD_LOCK(&accelerator_rocm_init_lock);
+
+    /* If already initialized, just exit */
+    if (true == accelerator_rocm_init_complete) {
+        goto out;
+    }
+
+    err = hipStreamCreate(&opal_accelerator_rocm_MemcpyStream);
+    if (hipSuccess != err) {
+        opal_output(0, "Could not create hipStream, err=%d %s\n",
+                err, hipGetErrorString(err));
+        goto out;
+    }
+
+    err = OPAL_SUCCESS;
+    opal_atomic_wmb();
+    accelerator_rocm_init_complete = true;
+out:
+    OPAL_THREAD_UNLOCK(&accelerator_rocm_init_lock);
+    return err;
+}
+
 static opal_accelerator_base_module_t* accelerator_rocm_init(void)
 {
+    OBJ_CONSTRUCT(&accelerator_rocm_init_lock, opal_mutex_t);
+    
     hipError_t err;
 
     if (opal_rocm_runtime_initialized) {
         return NULL;
     }
-
+    
     int count=0;
     err = hipGetDeviceCount(&count);
```

Comment:
The lazy init is a bit pointless if you have this check here. I also thought of a problem, it's going to be hard to ship builds containing rocm and cuda together because they would both show available if the header files/libraries can be found. Lazy init is a real pain and makes it hard to do any proper selection logic.

Reply:
Our line of thinking here is that the hipGetDeviceCount does not really cause any major issues, you are not yet consuming resources, but you are allowing the component to actually disqualify itself if there is no AMD GPU on the system. Its the hipStreamCreate that has a notable overhead that we would save with the lazy init.

Reply:
Makes sense if hipGetDeviceCount doesn't need any prior initialization. The reason cuda does a lazy init is because it requires a cuda context to be set before any of the API's are called. Not necessarily to save overhead from stream create

Reply:
There may be a way to see if there are any devices available and disqualify cuda in a similar way, it's a good line of thinking

Reply:
Hi William (@wckzhang),
I was wondering if the PR looks good to be merged in your view!
Please let me know.
Thank you!